frmts/gtiff/gt_overview.cpp

```diff
@@ -375,6 +375,10 @@ CPLErr GTIFFBuildOverviewsEx(const char *pszFilename, int nBands,
                 nBandFormat = SAMPLEFORMAT_INT;
                 break;
 
+            case GDT_Float16:
+                CPLAssert(false);
```

Comment:
The assertion is inappropriate, as this could happen. We should error out with a proper CPLError(CE_Failure, CPLE_NotSupported, ...) error. Same remark for CFloat16 below. Actually we could potentially support Float16. There's for example support for it in frmts/gtiff/gtiffoddbitsband.cpp